### PR TITLE
[SYCL][Graph] Implement graph and node properties

### PR DESCRIPTION
**Pull request by reble**
Adresses #104

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 1]
```diff
@@ -36,6 +36,38 @@ enum class graph_state {
...
+// Forward declaration
+class node;
+
+namespace property {
+namespace graph {
+
+class no_cycle_check : public ::sycl::detail::DataLessProperty<
+                           ::sycl::detail::GraphNoCycleCheck> {
+public:
+  no_cycle_check() = default;
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp:49]: Could we have a test that passes the `no_cycle_check` property to command_graph construction to check this is can be passed correctly? (test doesn't have to start trying to create cycles or anything)

**Review by EwanC — APPROVED** [review]
Not sure what changes are pending with this PR still marked "draft", but LGTM

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 2]
```diff
@@ -36,6 +36,38 @@ enum class graph_state {
...
+
+namespace node {
+
+class depends_on : public ::sycl::detail::PropertyWithData<
+                       ::sycl::detail::GraphNodeDependencies> {
+public:
+  template <typename... NodeTN> depends_on(NodeTN... nodes) : MDeps{nodes...} {}
+  const std::vector<::sycl::ext::oneapi::experimental::node> &
+  get_dependencies() {
+    return MDeps;
```
Comment (Bensuo) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp]:
```suggestion
  get_dependencies() const {
    return MDeps;
```

**Review by Bensuo — COMMENTED** [review]
LGTM, since this PR is a draft though are there more changes to be done?

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp (hunk 1 above)
Comment (reble) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp:49]: Makes sense. I've extended an existing test (empty node). Once implemented, we might want to extend the tests for cycle check.

**Review by Bensuo — APPROVED** [review]

**Pull request merged**